File: core/board_state.py
import numpy as np
import queue
from core.vector2 import Vector2
from core.action import ActionType

directions = [Vector2(1, 0), Vector2(-1, 0), Vector2(0, 1), Vector2(0, -1)]


class BoardState:

    def __init__(self, player_positions=[], player_walls=[], walls=np.zeros((8, 8), np.int8)):
        self.player_positions = player_positions
        self.player_walls = player_walls
        self.walls = walls
        self.distance_matrix = None
        self.debug_matrix = None

    def __repr__(self):
        output = ''
        for y in range(19):  # 9 cells + 8 inner walls + 2 outer walls
            for x in range(19):
                cell_char = ' '
                is_vert_wall = x % 2 == 0
                is_hori_wall = y % 2 == 0
                if y == 0 or y == 18:
                    cell_char = '―'
                elif x == 0 or x == 18:
                    cell_char = '|'
                elif not is_vert_wall and not is_hori_wall:  # is this a cell?
                    cell_x = int((x - 1) / 2)
                    cell_y = int((y - 1) / 2)
                    if self.is_cell_occupied(Vector2(cell_x, cell_y)):
                        cell_char = 'P'
                    elif self.distance_matrix is not None:
                        val = self.distance_matrix[cell_x][cell_y]
                        cell_char = ' ' if val is None else str(val % 10)
                elif is_vert_wall and is_hori_wall:  # Corners
                    cell_char = '·'
                elif is_vert_wall:
                    wall_pos_x = (x - 2) / 2
                    wall_pos_y = (y - 1) / 2
                    if self.is_path_blocked(Vector2(wall_pos_x, wall_pos_y), Vector2(1, 0)):
                        cell_char = '|'
                elif is_hori_wall:
                    wall_pos_x = (x - 1) / 2
                    wall_pos_y = (y - 2) / 2
                    if self.is_path_blocked(Vector2(wall_pos_x, wall_pos_y), Vector2(0, 1)):
                        cell_char = '―'
                output += cell_char
            output += '\n'
        return output

    def clone(self):
        return BoardState(list(self.player_positions), list(self.player_walls), np.copy(self.walls))

    def apply_action(self, action, player_index):
        new_board = self.clone()
        if action.type == ActionType.MOVE:
            new_board.player_positions[player_index] = action.new_pos
        elif action.type == ActionType.BLOCK:
            new_board.walls[action.block_pos.x][action.block_pos.y] = action.block_orientation
            new_board.player_walls[player_index] -= 1
        return new_board

    def switch(self):
        return BoardState(
            [Vector2(8, 8) - self.player_positions[0], Vector2(8, 8) - self.player_positions[1]],
            np.rot90(self.walls, 2))

    def is_cell_occupied(self, pos):
        return any(x == pos for x in self.player_positions)

    def is_path_blocked(self, cell, dir):
        orientation = 1 if dir.y == 0 else 2
        points = BoardState.get_wall_points(cell, dir)
        for point in points:
            if self.is_wall_index_in_bounds(point) and self.walls[point.x, point.y] == orientation:
                return True
        return False

    @staticmethod
    def get_wall_points(cell, dir):
        center = cell - Vector2(0.5, 0.5)
        cdir = dir / 2
        perp = Vector2(cdir.y, cdir.x)
        wall = center + cdir
        point1 = (wall + perp).as_int()
        point2 = (wall - perp).as_int()
        return (point1, point2)

    # For a given wall placement get which paths are blocked.
    @staticmethod
    def get_blocked_paths(wall, orientation):
        middle_offset = Vector2(0.5, 0.5)
        c_wall = wall + middle_offset
        direction = Vector2(0, 1) if orientation == 1 else Vector2(1, 0)
        c_dir = direction / 2
        perp = Vector2(c_dir.y, c_dir.x)
        return ((c_wall - perp - c_dir, c_wall - perp + c_dir), (c_wall + perp - c_dir, c_wall + perp + c_dir))

    def is_wall_index_in_bounds(self, i):
        return i.x >= 0 and i.y >= 0 and i.x < 8 and i.y < 8

    def is_cell_index_in_bounds(self, i):
        return i.x >= 0 and i.y >= 0 and i.x < 9 and i.y < 9

    def is_valid_wall(self, pos, orientation):
        if self.walls[pos.x, pos.y] != 0:
            return False
        shift_amount = Vector2(1, 0) if orientation == 1 else Vector2(0, 1)
        point_a = pos + shift_amount
        point_b = pos - shift_amount
        if self.is_wall_index_in_bounds(point_a) and self.walls[point_a.x, point_a.y] == orientation:
            return False
        if self.is_wall_index_in_bounds(point_b) and self.walls[point_b.x, point_b.y] == orientation:
            return False
        return True

    def get_distance_matrix(self, pos):
        matrix = np.full((9, 9), None)
        matrix[pos.x, pos.y] = 0
        q = queue.Queue()
        q.put(pos)
        while not q.empty():
            cell_pos = q.get()
            distance = matrix[cell_pos.x, cell_pos.y]
            for cell in self._get_accessible_adjacent_cells(cell_pos):
                if matrix[cell.x, cell.y] is None:
                    matrix[cell.x, cell.y] = distance + 1
                    q.put(cell)
        return matrix

    def get_distance_matrix_from_row(self, row):
        matrix = np.full((9, 9), None)
        q = queue.Queue()
        for x in range(9):
            matrix[x, row] = 0
            q.put(Vector2(x, row))
        while not q.empty():
            cell_pos = q.get()
            distance = matrix[cell_pos.x, cell_pos.y]
            for cell in self._get_accessible_adjacent_cells(cell_pos):
                if matrix[cell.x, cell.y] is None:
                    matrix[cell.x, cell.y] = distance + 1
                    q.put(cell)
        return matrix

    def get_deviation_matrix(self, dist, start):
        waves = queue.Queue()
        first_wave = queue.Queue()
        waves.put(first_wave)

        deviation_matrix = np.full((9, 9), None)
        deviation_matrix[start.x, start.y] = 0
        valid_moves = self.get_valid_moves(start, start)
        min_distance = -1
        for move in valid_moves:
            first_wave.put(move)
            move_distance = dist[move.x, move.y]
            if min_distance == -1 or move_distance < min_distance:
                min_distance = move_distance

        wave_count = 0
        while not waves.empty() and wave_count < 7:
            wave_count += 1
            next_wave = queue.Queue()
            current_wave = waves.get()
            while not current_wave.empty():
                cell = current_wave.get()
                distance = dist[cell.x, cell.y]
                deviation_matrix[cell.x, cell.y] = distance - min_distance
                for direction in directions:
                    new_position = cell + direction
                    if (self.is_cell_index_in_bounds(new_position)
                        and deviation_matrix[new_position.x, new_position.y] == None
                        and not self.is_path_blocked(cell, direction)):
                        next_wave.put(cell + direction)
            min_distance -= 1
            if not next_wave.empty():
                waves.put(next_wave)
        return deviation_matrix

    def get_valid_moves(self, from_pos, opp_pos):
        for pos in self._get_accessible_adjacent_cells(from_pos):
            if pos == opp_pos:
                for move in self.get_valid_moves(opp_pos, opp_pos):
                    yield move
            else:
                yield pos

    def _get_accessible_adjacent_cells(self, pos):
        for direction in directions:
            new_pos = pos + direction
            if not self.is_cell_index_in_bounds(new_pos):
                continue
            if not self.is_path_blocked(pos, direction):
                yield new_pos

File: c++/src/board_state.cc
#include "board_state.h"
#include <iostream>
#include <queue>
#include <memory>

using namespace std;

BoardState::BoardState() : walls_(Matrix(8, 8)) {
    player_positions_[0] = Vectori(4, 0);
    player_positions_[1] = Vectori(4, 8);
    player_wall_counts_[0] = 10;
    player_wall_counts_[1] = 10;
    distance_matrices_[0] = CalculateDistanceMatrix(8);
    distance_matrices_[1] = CalculateDistanceMatrix(0);
}

BoardState::BoardState(const BoardState& other) : 
    walls_(Matrix(other.walls_)),
    player_positions_(other.player_positions_),
    player_wall_counts_(other.player_wall_counts_),
    distance_matrices_(other.distance_matrices_) {
}

BoardState::BoardState(const BoardState& other, Action& action, int player_index) : 
    BoardState(other) {
        action.Apply(*this, player_index);
}

WallOrientation BoardState::GetWall(const Vectori& position) const {
    return static_cast<WallOrientation>(walls_[position]);
}

void BoardState::SetWall(const Vectori& position, WallOrientation value) {
    walls_[position] = value;
    distance_matrices_[0] = CalculateDistanceMatrix(8);
    distance_matrices_[1] = CalculateDistanceMatrix(0);
}

Vectori BoardState::GetPlayerPosition(int player_index) const {
    return player_positions_[player_index];
}

void BoardState::SetPlayerPosition(int player_index, const Vectori& position) {
    player_positions_[player_index] = position;
}

int BoardState::GetPlayerWallCount(int player_index) const {
    return player_wall_counts_[player_index];
}

void BoardState::SetPlayerWallCount(int player_index, int value) {
    player_wall_counts_[player_index] = value;
}

bool BoardState::IsCellOccupied(const Vectori& position) const {
    return (position == player_positions_[0] || position == player_positions_[1]);
}

bool BoardState::IsPathBlocked(const Vectori& cell, const Vectori& direction) const {
    auto orientation = direction.y == 0 ? 1 : 2;
    Vectori points[2];
    BoardState::GetWallPoints(cell, direction, points[0], points[1]);
    for (auto point : points) {
        if (IsWallIndexInBounds(point) && walls_[point] == orientation)
            return true;
    }
    return false;
}

vector<vector<Vectori>> BoardState::GetBlockedPaths(const Vectori& wall_position, int orientation) { 
    auto middle_offset = Vectord(0.5, 0.5);
    auto c_wall = wall_position + middle_offset;
    auto direction = orientation == 1 ? Vectori(1, 0) : Vectori(0, 1);
    auto c_dir = direction / 2;
    auto perp = Vectord(c_dir.y, c_dir.x);
    auto path1 = vector<Vectori> { c_wall - perp - c_dir, c_wall - perp + c_dir };
    auto path2 = vector<Vectori> { c_wall + perp - c_dir, c_wall + perp + c_dir };
    return vector<vector<Vectori>> { path1, path2 };
}

void BoardState::GetWallPoints(const Vectori& cell, const Vectori& direction, Vectori& point_1, Vectori& point_2) {
    auto center = cell - Vectord(0.5, 0.5);
    auto cdir = direction / 2;
    auto perp = Vectord(cdir.y, cdir.x);
    auto wall = center + cdir;
    point_1 = Vectori(wall + perp);
    point_2 = Vectori(wall - perp);
}

bool BoardState::IsWallIndexInBounds(const Vectori& cell) const {
    return cell.x >= 0 && cell.y >= 0 && cell.x < 8 && cell.y < 8;
}

bool BoardState::IsCellIndexInBounds(const Vectori& cell) const {
    return cell.x >= 0 && cell.y >= 0 && cell.x < 9 && cell.y < 9;
}

const Matrix& BoardState::GetDistanceMatrix(int player_index) const {
    return distance_matrices_[player_index];
}

Matrix BoardState::CalculateDistanceMatrix(int row) const {
    auto matrix = Matrix(9, 9, -1);
    auto cell_queue = queue<Vectori>();
    for (int x = 0; x < 9; x++) {
        auto cell = Vectori(x, row);
        matrix[cell] = 0;
        cell_queue.push(cell);
    }
    while (!cell_queue.empty()) {
        auto cell = cell_queue.front();
        cell_queue.pop();
        auto distance = matrix[cell];
        for (auto direction : directions) {
            auto adjacent_cell = cell + direction;
            if (IsCellIndexInBounds(adjacent_cell) && matrix[adjacent_cell] == -1 && !IsPathBlocked(cell, direction)) {
                matrix[adjacent_cell] = distance + 1;
                cell_queue.push(adjacent_cell);
            }
        }
    }
    return matrix;
}

Matrix BoardState::CalculateDeviationMatrix(const Matrix& distance_matrix, const Vectori& start_pos, const int max_waves) const {
    auto current_wave = vector<Vectori>();
    auto next_wave = vector<Vectori>();

    auto deviation_matrix = Matrix(9, 9, -1);
    deviation_matrix[start_pos] = 0;
    auto valid_moves = GetValidMoves(start_pos, start_pos);
    auto min_distance = -1;
    for (auto move : valid_moves) {
        next_wave.push_back(move);
        auto move_distance = distance_matrix[move];
        if (min_distance == -1 || move_distance < min_distance) {
            min_distance = move_distance;
        }
    }

    auto wave_count = 0;
    while (!next_wave.empty() && wave_count < max_waves) {
        wave_count += 1;
        current_wave.swap(next_wave);
        next_wave.clear();
        for (auto cell : current_wave) {
            auto distance = distance_matrix[cell];
            deviation_matrix[cell] = distance - min_distance;
            for (auto direction : directions) {
                auto new_position = cell + direction;
                if (IsCellIndexInBounds(new_position) 
                    && deviation_matrix[new_position] == -1 
                    && !IsPathBlocked(cell, direction)) {
                    next_wave.push_back(new_position);
                }
            }
        }
        min_distance -= 1;
    }
    return deviation_matrix;
}

vector<Vectori> BoardState::GetAccessibleAdjacentCells(const Vectori& cell) const {
    auto cells = vector<Vectori>();
    for (auto direction : directions) {
        auto new_cell = cell + direction;
        if (IsCellIndexInBounds(new_cell) && !IsPathBlocked(cell, direction)) {
            cells.push_back(new_cell);
        }
    }
    return cells;
}

vector<Vectori> BoardState::GetValidMoves(const Vectori& from_pos, const Vectori& opp_pos) const {
    auto result = vector<Vectori>();
    for (auto position : GetAccessibleAdjacentCells(from_pos)) {
        if (position == opp_pos) {
            for (auto jump_position : GetValidMoves(opp_pos, opp_pos)) {
                if (jump_position != position) {
                    result.push_back(jump_position);
                }
            }
        }
        else {
            result.push_back(position);
        }
    }
    return result;
}

std::vector<Vectori> BoardState::GetValidMoves(int player_index) const {
    return GetValidMoves(GetPlayerPosition(player_index), GetPlayerPosition(1 - player_index));
}

ostream &operator<< (ostream &os, const BoardState &bs) {
    for (int y = 18; y >= 0; y--) { // 9 cells + 8 inner walls + 2 outer walls
        for (int x = 0; x < 19; x++) {
            auto cell_char = " ";
            auto is_vert_wall = x % 2 == 0;
            auto is_hori_wall = y % 2 == 0;
            if (y == 0 || y == 18)
                cell_char = "―";
            else if (x == 0 || x == 18)
                cell_char = "|";
            else if (!is_vert_wall && !is_hori_wall) { // is this a cell?
                auto cell_x = int((x - 1) / 2);
                auto cell_y = int((y - 1) / 2);
                auto cell = Vectori(cell_x, cell_y);
                if (bs.GetPlayerPosition(0) == cell)
                    cell_char = "1";
                else if (bs.GetPlayerPosition(1) == cell)
                    cell_char = "2";
                // else if (self.distance_matrix is not None) {
                //     val = self.distance_matrix[cell_x][cell_y]
                //     cell_char = ' ' if val is None else str(val % 10)
                // }
            }
            else if (is_vert_wall && is_hori_wall) { // Corners
                cell_char = "·";
            }
            else if (is_vert_wall) {
                auto wall_pos_x = (x - 2) / 2;
                auto wall_pos_y = (y - 1) / 2;
                if (bs.IsPathBlocked(Vectori(wall_pos_x, wall_pos_y), Vectori(1, 0)))
                    cell_char = "|";
            }
            else if (is_hori_wall) {
                auto wall_pos_x = (x - 1) / 2;
                auto wall_pos_y = (y - 2) / 2;
                if (bs.IsPathBlocked(Vectori(wall_pos_x, wall_pos_y), Vectori(0, 1)))
                    cell_char = "―";
            }
            os << cell_char;
        }            
        os << endl;
    }        
    return os;
};